firedrake/variational_solver.py

```diff
@@ -77,7 +77,7 @@ def __init__(self, F, u, bcs=None, J=None,
             raise TypeError("Provided solution is a '%s', not a Function" % type(self.u).__name__)
         # Use the user-provided Jacobian. If none is provided, derive
         # the Jacobian from the residual.
-        self.J = J or ufl_expr.derivative(F, u)
+        self.J = J or ufl.lhs(ufl_expr.derivative(F, u))
```

Comment:
Given that the Jacobian is a 2-form, `ufl.lhs` won't do anything except expanding derivatives. So `expand_derivatives` is probably what you want here.